### ISSUE TITLE
fclose(STDIN); gives InvalidPassByReference

### ISSUE THREAD

**Issue by thomasbley** [vimeo/psalm#7505]
https://psalm.dev/r/1d03e9adab

fclose(STDIN); should be possible in the same way as fgets(STDIN).

https://www.php.net/manual/en/features.commandline.io-streams.php

Tested with PHP 8.0.4 and Psalm 4.19.0 (problem also in previous versions).

**Comment by AndrolGenhald** [comment]
I'm not really sure what we want to do here. The rule exists because passing a literal or const to a function wanting a reference is an [error](https://3v4l.org/6mASI), but apparently PHP allows using const resources without issue, even though they're not really consts (they can be closed...).

Tested locally since you can't use files on 3v4l:
```
php > define("SOME_CONST", fopen("test", "r"));
php > echo fgets(SOME_CONST);
test
php > fclose(SOME_CONST);
php > echo fgets(SOME_CONST);
PHP Warning:  Uncaught TypeError: fgets(): supplied resource is not a valid stream resource in php shell code:1
Stack trace:
#0 php shell code(1): fgets()
#1 {main}
  thrown in php shell code on line 1
```

This seems like poor behavior from PHP. I think maybe we should make sure all functions that take file handles use references, and people can work around it like [this](https://psalm.dev/r/6f3b20d44a).

I guess how you'd expect it to behave depends on whether you see the file handle as just a number or a `resource`, but PHP decided to add the special `resource` type and handle it differently, so it seems very strange that it would be allowed to be a constant.

**Comment by weirdan** [comment]
`fclose()` actually does not accept its argument by reference. I'm pretty sure it was specified as reference _just_ to use `@param-out`.

Ref: 8ccd4860e3, cc: @muglug

**Comment by AndrolGenhald** [comment]
> `fclose()` actually does not accept its argument by reference. I'm pretty sure it was specified as reference _just_ to use `@param-out`.
> 
> Ref: [8ccd486](https://github.com/vimeo/psalm/commit/8ccd4860e369cd71b4dbfc3bc4ebf92d41004e6f), cc: @muglug

True, but given that the constant isn't actually a constant it feels weird to me to allow it.

**Comment by weirdan** [comment]
> Should I go edit the callmap to change all resource arguments to references

Please don't. That would be misleading.

> it seems very strange that it would be allowed to be a constant.

Oh, even mutable objects are allowed as constants nowadays (https://3v4l.org/VLG4Y#v8.1.2). So resources as constants don't strike me as weird at all.

I believe the best (but certainly not the easiest) approach would be to ditch 'closed-resource' in favour of `resource<'closed'>`. This way it would be apparent that the type itself doesn't change, but its internal state does.

**Comment by VincentLanglet** [comment]
> `fclose()` actually does not accept its argument by reference. I'm pretty sure it was specified as reference _just_ to use `@param-out`.
> 
> Ref: [8ccd486](https://github.com/vimeo/psalm/commit/8ccd4860e369cd71b4dbfc3bc4ebf92d41004e6f), cc: @muglug

Wouldn't be able for psalm to automatically suppress the InvalidPassByReference for the fclose method ?

Because the error is misleading. `fclose($this->getFile())` is perfectly valid.
